core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/WorkerBase.scala

```diff
@@ -79,14 +87,26 @@ abstract class WorkerBase(identifier: ActorVirtualIdentity) extends WorkflowActo
     context.parent ! ReportState(WorkerState.Running)
   }
 
+//  val afterPause = () =>{
+//    // if dp thread is blocking on waiting for input tuples:
+//    if (dataProcessor.isQueueEmpty) {
+//      // insert dummy batch to unblock dp thread
+//      dataProcessor.appendElement(DummyInput())
+//    }
+//    context.become(pausing)
+//  }
+//
+//  val finishPause = () =>{
+//    logger.info(s"received Excution Pause message")
+//    onPaused()
+//    context.become(paused)
+//    unstashAll()
+//  }
+
   def onPausing(): Unit = {
-    //messagingManager.pauseDataSending()
-    pauseManager.pause()
-    // if dp thread is blocking on waiting for input tuples:
-    if (dataProcessor.isQueueEmpty) {
-      // insert dummy batch to unblock dp thread
-      dataProcessor.appendElement(DummyInput())
-    }
+    // messagingManager.pauseDataSending()
+    // pauseManager.pause()
+    promiseManager.execute(PromiseInvocation(null, WorkerPause()))
```

Comment:
shouldn't this go through ControlInputPort?

Reply:
In order to not involve any change of principal and controller, I have to do it this way. After we refactored the controller and all messages are sent through the messaging layer, we won't have that line of code.

Reply:
I dont get it. lets discuss